File: tools/knowledge/probe.py
#!/usr/bin/env python3
"""Retrieval probe: routing, answer recall and language bias, measured.

    engine/.venv/bin/python tools/knowledge/probe.py               # keywords, dense, hybrid
    engine/.venv/bin/python tools/knowledge/probe.py --calibrate   # the dense similarity floor
    engine/.venv/bin/python tools/knowledge/probe.py "any question"
    engine/.venv/bin/python tools/knowledge/probe.py --graph       # the full agentic pipeline

OPENAI_API_KEY comes from the environment or engine/.env. Without it only the
keyword ranking can be measured. --graph runs the twin's graph up to the
evidence it would answer from (route, rewrite, retrieve, grade), so it spends
gpt-4o-mini tokens -- roughly 150K for the whole probe -- but writes no answers.
"""
from __future__ import annotations

import asyncio
import json
import re

import openai
import statistics
import sys
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT / "engine"))

import config  # noqa: E402  (loads engine/.env)
from rag.retriever import KnowledgeIndex  # noqa: E402

EVAL_FILE = ROOT / "engine" / "eval" / "retrieval.json"
EVAL = json.loads(EVAL_FILE.read_text(encoding="utf8"))
ROUTING = [(case["question"], case["sourcePrefix"]) for case in EVAL["routing"]]
NEEDLES = [
    (case["question"], re.compile(case["answerPattern"])) for case in EVAL["answerRecall"]
]
OPEN = EVAL["openQuestions"]
OFF_TOPIC = EVAL["offTopic"]
STARTERS = EVAL["starters"]
RECALL_FLOOR = EVAL["recallFloor"]


def pct(part: int, whole: int) -> str:
    return f"{100 * part / (whole or 1):.0f}%"


def line(passage: dict) -> str:
    where = f"p{passage['page']}" if passage.get("page") else "md"
    text = " ".join(passage["text"].split())[:80]
    return f"[{passage['lang']}] {passage['source'][:34]:34} {where:5} {text}…"


async def report(index: KnowledgeIndex, modes: list[str]) -> int:
    main = modes[-1]

    print(f"\nROUTING — expected document ranked first ({main})")
    routed = 0
    for question, expected in ROUTING:
        hits = await index.retrieve(question, mode=main)
        got = hits[0]["source"] if hits else "nothing"
        ok = got.startswith(expected)
        routed += ok
        print(f"  {'PASS' if ok else 'FAIL'}  {question}{'' if ok else f'  → {got}'}")

    print("\nANSWER RECALL @4 — English question, answer written only in French")
    print("  " + "".join(f"{mode:9}" for mode in modes) + "question")
    recall = dict.fromkeys(modes, 0)
    for question, pattern in NEEDLES:
        marks = []
        for mode in modes:
            hits = await index.retrieve(question, mode=mode)
            ok = any(pattern.search(hit["text"]) for hit in hits)
            recall[mode] += ok
            marks.append(f"{'✓' if ok else '·':9}")
        print("  " + "".join(marks) + question)
    print("  " + "".join(f"{f'{recall[m]}/{len(NEEDLES)}':9}" for m in modes) + "total")

    print("\nLANGUAGE — English share of passages retrieved for open questions")
    english = sum(chunk["lang"] == "en" for chunk in index.chunks)
    print(f"  {'corpus':9}{pct(english, index.size)}")
    for mode in modes:
        en = total = 0
        for question in OPEN:
            for hit in await index.retrieve(question, mode=mode):
                total += 1
                en += hit["lang"] == "en"
        print(f"  {mode:9}{pct(en, total)}  ({en}/{total})")

    floor = RECALL_FLOOR if index.embed_model else 0
    print(f"\n{routed}/{len(ROUTING)} routed · {main} recall {recall[main]}/{len(NEEDLES)} (floor {floor})")
    return 0 if routed == len(ROUTING) and recall[main] >= floor else 1


async def calibrate(index: KnowledgeIndex) -> int:
    if not index.embed_model:
        print("dense retrieval is off: nothing to calibrate")
        return 1

    async def top(question: str) -> float:
        nearest = await index.nearest(question, 1)
        return nearest[0]["similarity"] if nearest else 0.0

    print("\nOFF-TOPIC — best similarity (these should retrieve nothing)")
    noise = 0.0
    for question in OFF_TOPIC:
        similarity = await top(question)
        noise = max(noise, similarity)
        print(f"  {similarity:.3f}  {question}")

    print("\nANSWERS — similarity of the passage that actually answers")
    signal = float("inf")
    for question, pattern in NEEDLES:
        ranked = await index.nearest(question, index.size)
        rank = next((i for i, hit in enumerate(ranked) if pattern.search(hit["text"])), -1)
        similarity = ranked[rank]["similarity"] if rank >= 0 else 0.0
        signal = min(signal, similarity)
        print(f"  {similarity:.3f}  rank {rank + 1:3}  {question}")

    print("\nOPEN — best similarity")
    for question in OPEN:
        print(f"  {await top(question):.3f}  {question}")

    print(f"\nnoise ceiling {noise:.3f} · weakest answer {signal:.3f}")
    if noise < signal:
        print(f"clean gap: a floor of ~{(noise + signal) / 2:.2f} drops every off-topic turn and keeps every answer")
    else:
        print("no clean gap: some off-topic turn scores as high as a real answer")
    return 0


async def graph_report(index: KnowledgeIndex, hybrid_mode: str) -> int:
    from langchain_core.callbacks import UsageMetadataCallbackHandler

    from agent.graph import build_graph

    if not config.OPENAI_API_KEY:
        print("--graph needs OPENAI_API_KEY")
        return 1
    # Generous retries: questions go back to back into a tokens-per-minute
    # limit, and the SDK's backoff paces them. A daily request cap cannot be
    # waited out, so that stops the run with whatever was measured so far.
    twin = build_graph(
        index,
        model=config.MODEL,
        api_key=config.OPENAI_API_KEY,
        base_url=config.OPENAI_BASE_URL,
        max_retries=8,
        answer=False,
    )
    latencies: list[float] = []
    tokens: list[int] = []

    async def run(question: str) -> dict:
        usage = UsageMetadataCallbackHandler()
        started = time.perf_counter()
        state = await twin.ainvoke({"message": question, "history": []}, config={"callbacks": [usage]})
        latencies.append(time.perf_counter() - started)
        tokens.append(sum(m.get("total_tokens", 0) for m in usage.usage_metadata.values()))
        return state

    def where(state: dict) -> str:
        evidence = state.get("evidence", [])
        kept = sum(not p.get("neighbour") for p in evidence)
        passes = state.get("attempt", 0)
        return f"{state['route']}, {kept} kept + {len(evidence) - kept} adjacent, {passes} grading pass(es)"

    recall = {"hybrid": 0, "kept": 0, "graph": 0}
    counts: dict = {}
    try:
        await graph_sections(index, hybrid_mode, run, where, recall, counts)
    except openai.RateLimitError as error:
        detail = error.body.get("message", str(error)) if isinstance(error.body, dict) else str(error)
        print(f"\nSTOPPED by the {config.MODEL} rate limit: {detail}")
    routed, quiet = counts.get("routed", 0), counts.get("quiet", 0)
    if latencies:
        ordered = sorted(latencies)
        print(
            f"\nbefore the first token: median {statistics.median(ordered):.2f}s, "
            f"p90 {ordered[int(0.9 * (len(ordered) - 1))]:.2f}s, max {ordered[-1]:.2f}s over {len(ordered)} questions "
            f"(route + retrieve + grade; rate-limit waits included)"
        )
        print(f"tokens before answering: median {statistics.median(tokens):.0f}, max {max(tokens)} per question")
    print(
        f"\n{routed}/{len(ROUTING)} routed · graph recall {recall['graph']}/{len(NEEDLES)} "
        f"(floor {RECALL_FLOOR}) · {quiet}/{len(OFF_TOPIC)} off-topic quiet"
    )
    return 0 if routed == len(ROUTING) and recall["graph"] >= RECALL_FLOOR else 1


async def graph_sections(index, hybrid_mode, run, where, recall: dict, counts: dict) -> None:
    print(f"\nGRAPH ROUTING — expected document first in the evidence ({config.MODEL})")
    routed = 0
    for question, expected in ROUTING:
        state = await run(question)
        evidence = state.get("evidence", [])
        got = evidence[0]["source"] if evidence else "nothing"
        ok = got.startswith(expected)
        routed += ok
        print(f"  {'PASS' if ok else 'FAIL'}  {question}  [{where(state)}]{'' if ok else f'  → {got}'}")
    counts["routed"] = routed

    print("\nGRAPH ANSWER RECALL — English question, answer written only in French")
    print(f"  {'hybrid@4':10}{'kept':6}{'+adjacent':11}question")
    for question, pattern in NEEDLES:
        baseline = any(
            pattern.search(hit["text"])
            for hit in await index.retrieve(question, mode=hybrid_mode)
        )
        state = await run(question)
        evidence = state.get("evidence", [])
        kept = any(pattern.search(p["text"]) for p in evidence if not p.get("neighbour"))
        full = any(pattern.search(p["text"]) for p in evidence)
        for name, ok in (("hybrid", baseline), ("kept", kept), ("graph", full)):
            recall[name] += ok
        marks = f"  {'✓' if baseline else '·':10}{'✓' if kept else '·':6}{'✓' if full else '·':11}"
        print(f"{marks}{question}  [{where(state)}]")
    totals = {name: f"{count}/{len(NEEDLES)}" for name, count in recall.items()}
    print(f"  {totals['hybrid']:10}{totals['kept']:6}{totals['graph']:11}total")

    print("\nGRAPH LANGUAGE — English share of the evidence for open questions")
    english = sum(chunk["lang"] == "en" for chunk in index.chunks)
    en = total = 0
    for question in OPEN:
        for passage in (await run(question)).get("evidence", []):
            total += 1
            en += passage["lang"] == "en"
    print(f"  corpus {pct(english, index.size)} · graph {pct(en, total)}  ({en}/{total})")

    print("\nGRAPH OFF-TOPIC — should cite nothing")
    quiet = counts["quiet"] = 0
    for question in OFF_TOPIC:
        state = await run(question)
        ok = not state.get("evidence")
        quiet += ok
        print(f"  {'PASS' if ok else 'FAIL'}  {question}  [{where(state)}]")
        counts["quiet"] = quiet

    print("\nSTARTERS — the studio's suggested questions")
    for question in STARTERS:
        state = await run(question)
        labels = list(dict.fromkeys(p["label"] for p in state.get("evidence", [])))
        print(f"  {question}  [{where(state)}]  sources: {', '.join(labels) or 'none'}")


async def adhoc(index: KnowledgeIndex, modes: list[str], question: str) -> int:
    for mode in modes:
        print(f"\n{mode}")
        for hit in await index.retrieve(question, mode=mode):
            print(f"  {line(hit)}")
    if index.embed_model:
        print("\nnearest by embedding")
        for hit in await index.nearest(question, 5):
            text = " ".join(hit["text"].split())[:80]
            print(f"  {hit['similarity']:.3f} [{hit['lang']}] {hit['source'][:34]:34} {text}…")
    return 0


async def main(argv: list[str]) -> int:
    index = KnowledgeIndex(
        config.KNOWLEDGE_DIR,
        openai_api_key=config.OPENAI_API_KEY,
        openai_base_url=config.OPENAI_BASE_URL,
    )
    dense = f"{index.embed_provider}:{index.embed_model}" if index.embed_model else f"off ({index.dense_inactive_reason})"
    print(f"index: {index.size} passages · dense: {dense}")
    modes = ["lexical", "dense", "hybrid"] if index.embed_model else ["lexical"]
    # one index behind every mode, so each question is embedded once
    if argv[:1] == ["--calibrate"]:
        return await calibrate(index)
    if argv[:1] == ["--graph"]:
        return await graph_report(index, "hybrid" if index.embed_model else "lexical")
    if argv:
        return await adhoc(index, modes, " ".join(argv))
    return await report(index, modes)


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main(sys.argv[1:])))

File: tools/knowledge/probe.mjs
/**
 * Retrieval probe: routing, answer recall, and language bias.
 *
 *   node --env-file-if-exists=engine/.env tools/knowledge/probe.mjs
 *   node --env-file-if-exists=engine/.env tools/knowledge/probe.mjs --calibrate
 *   node --env-file-if-exists=engine/.env tools/knowledge/probe.mjs "any question"
 *
 * The env file supplies OPENAI_API_KEY when the index holds OpenAI vectors.
 * Without it the probe measures lexical retrieval only.
 */
import { readFileSync } from "node:fs";
import { createRetriever } from "../../engine/rag/retrieve.mjs";

// Routing: the document each question should reach first. Asserted on the top
// hit, not "somewhere in the top 4": the thesis alone is three quarters of the
// index, so the weaker test passes by accident.
const ROUTING = [
  ["What is your thesis about?", "These_VF"],
  ["Tell me about world models and ontologies", "These_VF"],
  ["What did you do at SaaSOffice?", "SaaSOffice_work"],
  ["What did you build during your apprenticeship at Amadeus?", "Rapport_REX"],
  ["How does Member Insight predict why a member is calling?", "member-insight"],
  ["How does the configuration chatbot use function calling?", "ALMS-V2"],
  ["Qu'est-ce que tu as fait chez SaaSOffice ?", "SaaSOffice_work"],
  ["decision intelligence", "These_VF"],
];

// Answer recall: English questions whose answer is written only in French. The
// pattern matches the sentence that actually answers, and each was checked to
// have no English counterpart anywhere in the corpus -- so a pass means the
// French prose was reached, not an English fragment near it.
const NEEDLES = [
  ["How did the SaaSOffice chatbot classify the people it talked to?", /Fournisseur|Candidat/],
  ["How did the SaaSOffice chatbot work when you arrived?", /appel consécutifs d'agents|sous-tâches/],
  ["How many employees did SaaSOffice have when you joined?", /quatre personnes|cinquantaine de salariés/],
  ["Why did your first configuration chatbot give wrong answers?", /pages Confluence avaient été écrites|doc drift/],
  ["What did demoing prototypes to clients teach you?", /annoncer une limite/],
  ["What changes for a GenAI developer when a prototype becomes a product?", /la difficulté se déplace|faisabilité technique/],
  ["How do you think about technical debt?", /dette technique/],
  ["Do people have to choose between transparency and performance in AI decisions?", /positivement\s+corrélées|transparence contre performance/],
];

// Language bias: open questions. The English share of what they retrieve should
// sit near the corpus share, not far above it.
const OPEN = [
  "What is your thesis about?",
  "What did you learn during your apprenticeship?",
  "What surprised you most while building AI at Amadeus?",
  "How do you evaluate AI systems?",
  "Tell me about world models",
  "What is decision intelligence?",
];

// Conversational and off-topic turns: nothing should be retrieved for these,
// which is what sets the dense similarity floor from below.
const OFF_TOPIC = [
  "hello",
  "thanks!",
  "what's the weather like today?",
  "tell me a joke",
  "who won the football last night?",
  "zzzz",
];

// Hybrid answer recall when text-embedding-3-large and the 1:2 fusion were
// chosen (dense alone scored 6/8, keywords alone 1/8). Falling below it fails
// the probe: something in retrieval regressed.
const RECALL_FLOOR = 5;

const retriever = createRetriever();
const modes = retriever.embedModel ? ["lexical", "dense", "hybrid"] : ["lexical"];
const main = modes.at(-1);
const corpus = JSON.parse(
  readFileSync(new URL("../../engine/knowledge/index.json", import.meta.url), "utf8"),
).chunks;

const pct = (part, whole) => `${((100 * part) / (whole || 1)).toFixed(0)}%`;
const line = (hit) =>
  `[${hit.lang}] ${hit.source.slice(0, 34).padEnd(34)} ${(hit.page ? `p${hit.page}` : "md").padEnd(5)} ${hit.text.slice(0, 80).replace(/\s+/g, " ")}…`;

console.log(
  `index: ${retriever.size} passages · dense: ${
    retriever.embedModel
      ? `${retriever.embedProvider}:${retriever.embedModel}`
      : `off (${retriever.denseInactiveReason})`
  }`,
);

async function report() {
  console.log(`\nROUTING — expected document ranked first (${main})`);
  let routed = 0;
  for (const [question, expected] of ROUTING) {
    const [hit] = await retriever.retrieve(question, { mode: main });
    const ok = (hit?.source ?? "").startsWith(expected);
    if (ok) routed++;
    console.log(`  ${ok ? "PASS" : "FAIL"}  ${question}${ok ? "" : `  → ${hit?.source ?? "nothing"}`}`);
  }

  console.log("\nANSWER RECALL @4 — English question, answer written only in French");
  console.log(`  ${modes.map((mode) => mode.padEnd(9)).join("")}question`);
  const recall = Object.fromEntries(modes.map((mode) => [mode, 0]));
  for (const [question, pattern] of NEEDLES) {
    const marks = [];
    for (const mode of modes) {
      const hits = await retriever.retrieve(question, { mode });
      const ok = hits.some((hit) => pattern.test(hit.text));
      if (ok) recall[mode]++;
      marks.push((ok ? "✓" : "·").padEnd(9));
    }
    console.log(`  ${marks.join("")}${question}`);
  }
  console.log(
    `  ${modes.map((mode) => `${recall[mode]}/${NEEDLES.length}`.padEnd(9)).join("")}total`,
  );

  console.log("\nLANGUAGE — English share of passages retrieved for open questions");
  const english = corpus.filter((chunk) => chunk.lang === "en").length;
  console.log(`  ${"corpus".padEnd(9)}${pct(english, corpus.length)}`);
  for (const mode of modes) {
    let en = 0;
    let all = 0;
    for (const question of OPEN)
      for (const hit of await retriever.retrieve(question, { mode })) {
        all++;
        if (hit.lang === "en") en++;
      }
    console.log(`  ${mode.padEnd(9)}${pct(en, all)}  (${en}/${all})`);
  }

  const floor = retriever.embedModel ? RECALL_FLOOR : 0;
  if (routed < ROUTING.length || recall[main] < floor) process.exitCode = 1;
  console.log(
    `\n${routed}/${ROUTING.length} routed · ${main} recall ${recall[main]}/${NEEDLES.length} (floor ${floor})`,
  );
}

async function calibrate() {
  if (!retriever.embedModel) {
    console.log("dense retrieval is off: nothing to calibrate");
    return;
  }
  const top = async (question) => (await retriever.nearest(question, 1))[0]?.similarity ?? 0;

  console.log("\nOFF-TOPIC — best similarity (these should retrieve nothing)");
  let noise = 0;
  for (const question of OFF_TOPIC) {
    const similarity = await top(question);
    noise = Math.max(noise, similarity);
    console.log(`  ${similarity.toFixed(3)}  ${question}`);
  }

  console.log("\nANSWERS — similarity of the passage that actually answers");
  let signal = Infinity;
  for (const [question, pattern] of NEEDLES) {
    const ranked = await retriever.nearest(question, retriever.size);
    const rank = ranked.findIndex((hit) => pattern.test(hit.text));
    const similarity = rank >= 0 ? ranked[rank].similarity : 0;
    signal = Math.min(signal, similarity);
    console.log(`  ${similarity.toFixed(3)}  rank ${String(rank + 1).padStart(3)}  ${question}`);
  }

  console.log("\nOPEN — best similarity");
  for (const question of OPEN) console.log(`  ${(await top(question)).toFixed(3)}  ${question}`);

  console.log(`\nnoise ceiling ${noise.toFixed(3)} · weakest answer ${signal.toFixed(3)}`);
  console.log(
    noise < signal
      ? `clean gap: a floor of ~${((noise + signal) / 2).toFixed(2)} drops every off-topic turn and keeps every answer`
      : "no clean gap: some off-topic turn scores as high as a real answer",
  );
}

async function adhoc(question) {
  for (const mode of modes) {
    console.log(`\n${mode}`);
    for (const hit of await retriever.retrieve(question, { mode })) console.log(`  ${line(hit)}`);
  }
  if (retriever.embedModel) {
    console.log("\nnearest by embedding");
    for (const hit of await retriever.nearest(question, 5))
      console.log(`  ${hit.similarity.toFixed(3)} ${line(hit)}`);
  }
}

const args = process.argv.slice(2);
if (args[0] === "--calibrate") await calibrate();
else if (args.length) await adhoc(args.join(" "));
else await report();
